### PR TITLE
Add support for favourites and a bunch of other cool stuff

### LINKED ISSUE #8: Add Favourites Support
Hello,

It would be great if the Brisk Menu could have a favorites menu and an option to be able to right click on any program and add it to that favorites menu,

Thank you,

EastAPOLO

### PR DESCRIPTION
**Pull request by cybre**
Things added in this commit:
  - Favourites backend
  - Context menu API
  - DesktopAppInfo actions

This resolves #8

### CODE REVIEW

File: src/backend/favourites/favourites-backend.c [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+        gstrv *old = NULL;
+        autofree(gstrv) *new = NULL;
+        gsize size;
+
+        const gchar *item_id = g_object_get_data(G_OBJECT(action), "__brisk_item_id");
+
+        old = g_settings_get_strv(self->settings, "favourites");
+
+        for (size = 0; old[size] != NULL; size++)
+                ;
```
Comment (ikeydoherty) [inline src/backend/favourites/favourites-backend.c]: Eh?

File: src/backend/favourites/favourites-backend.c [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+}
+
+/**
+ * brisk_favourites_backend_init:
+ *
+ * Handle construction of the BriskFavouritesBackend
+ */
+static void brisk_favourites_backend_init(__brisk_unused__ BriskFavouritesBackend *self)
+{
+        self->settings = g_settings_new("com.solus-project.brisk-menu");
```
Comment (ikeydoherty) [inline src/backend/favourites/favourites-backend.c]: This guy is being leaked

File: src/frontend/entry-button.c [hunk 3]
```diff
@@ -216,6 +215,9 @@ static void brisk_menu_entry_button_init(BriskMenuEntryButton *self)
         gtk_misc_set_alignment(GTK_MISC(self->label), 0.0, 0.5);
         G_GNUC_END_IGNORE_DEPRECATIONS
 
+        menu = gtk_menu_new();
```
Comment (ikeydoherty) [inline src/frontend/entry-button.c]: Can probably just directly set `self->menu = gtk_menu_new()` here, no need for the second `menu` pointer really

File: src/frontend/entry-button.c [hunk 4]
```diff
@@ -298,10 +300,73 @@ static void brisk_menu_entry_drag_data(GtkWidget *widget, __brisk_unused__ GdkDr
         gtk_selection_data_set_uris(data, (gchar **)uris);
 }
 
-static void brisk_menu_entry_clicked(GtkButton *widget)
+static void build_menu(__brisk_unused__ gpointer key, gpointer value, gpointer user_data)
+{
+        BriskBackend *backend = BRISK_BACKEND(value);
+        BriskMenuEntryButton *self = BRISK_MENU_ENTRY_BUTTON(user_data);
+        GSList *actions = brisk_backend_get_item_actions(backend, self->item);
```
Comment (ikeydoherty) [inline src/frontend/entry-button.c]: I'm fairly sure these things are leaking all over the place

File: src/frontend/entry-button.c [hunk 5]
```diff
@@ -298,10 +300,73 @@ static void brisk_menu_entry_drag_data(GtkWidget *widget, __brisk_unused__ GdkDr
...
+        }
+}
+
+static gboolean brisk_menu_entry_button_release_event(GtkWidget *widget,
+                                                      GdkEventButton *event_button)
 {
         BriskMenuEntryButton *self = BRISK_MENU_ENTRY_BUTTON(widget);
-        brisk_menu_entry_button_launch(self);
+
+        if (event_button->button == GDK_BUTTON_PRIMARY) {
```
Comment (ikeydoherty) [inline src/frontend/entry-button.c]: I don't like how this flows, the `else if` is just unnecessary. Handle the primary, then if the button **doesn't match** the secondary, just return in these spots on the parent method. Saves a whole bunch of indentation

File: src/frontend/entry-button.c [hunk 6]
```diff
@@ -298,10 +300,73 @@ static void brisk_menu_entry_drag_data(GtkWidget *widget, __brisk_unused__ GdkDr
...
+                for (elem = kids; elem; elem = elem->next) {
+                        GtkWidget *row = elem->data;
+                        gtk_widget_destroy(row);
+                }
+                g_list_free(kids);
+
+                BriskMenuWindow *window = BRISK_MENU_WINDOW(gtk_widget_get_toplevel(widget));
+                g_hash_table_foreach(window->backends, build_menu, widget);
+
+                gtk_menu_popup(GTK_MENU(self->menu),
```
Comment (ikeydoherty) [inline src/frontend/entry-button.c]: With the fixed indentation `clang-format` will be more forgiving for this indent issue here.

File: src/lib/key-binder.c [hunk 7]
```diff
@@ -186,7 +186,11 @@ static GdkFilterReturn brisk_key_binder_filter(GdkXEvent *xevent, GdkEvent *even
                         /* when breaking the shortcut sequence, send the event up the window
                          * hierarchy in case it's part of a different shortcut sequence
                          * (e.g. <Mod4>a) */
-                        XSendEvent(display, GDK_WINDOW_XID(self->root_window), TRUE, KeyPressMask | KeyReleaseMask, xev);
+                        XSendEvent(display,
```
Comment (ikeydoherty) [inline src/lib/key-binder.c]: I assume this was clang-format acting the bollocks here, not you changing it?

File: src/backend/favourites/favourites-section.c [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        autofree(gstrv) *favourites = NULL;
+
+        BriskFavouritesSection *self = BRISK_FAVOURITES_SECTION(section);
+        favourites = g_settings_get_strv(self->settings, "favourites");
+
+        if (!favourites) {
+                return FALSE;
+        }
+
+        return g_strv_contains((const gchar *const *)favourites, brisk_item_get_id(item));
```
Comment (ikeydoherty) [inline src/backend/favourites/favourites-section.c]: I'm worried about the efficiency here tbqh. I can optimize it post-merge though.

File: src/backend/favourites/favourites-backend.c [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+
+static const gchar *brisk_favourites_backend_get_display_name(
+    __brisk_unused__ BriskBackend *backend)
+{
+        return _("Favourites");
+}
+
+static GSList *brisk_favourites_backend_get_item_actions(BriskBackend *backend, BriskItem *item)
+{
+        GSList *list = NULL;
```
Comment (ikeydoherty) [inline src/backend/favourites/favourites-backend.c]: This list is never actually freed anywhere and the actions leak

File: src/backend/backend.h [hunk 10]
```diff
@@ -24,9 +24,8 @@ typedef struct _BriskBackendClass BriskBackendClass;
  * Flags to indicate the support level of a given backend
  */
 typedef enum {
-        BRISK_BACKEND_FAVOURITES = 1 << 0, /**<Supports "favourites" management */
-        BRISK_BACKEND_KEYBOARD = 1 << 1,   /**<Supports keyboard shortcuts */
-        BRISK_BACKEND_SOURCE = 1 << 2,     /**<Provides data which must be loaded */
+        BRISK_BACKEND_KEYBOARD = 1 << 0, /**<Supports keyboard shortcuts */
+        BRISK_BACKEND_SOURCE = 1 << 1,   /**<Provides data which must be loaded */
```
Comment (ikeydoherty) [inline src/backend/backend.h]: Good call dropping the FAVOURITES notion :)

File: src/backend/backend.c [hunk 11]
```diff
@@ -252,49 +310,15 @@ const gchar *brisk_backend_get_display_name(BriskBackend *backend)
...
- * Unpin a previously pinned item.
- */
-gboolean brisk_backend_unpin_item(BriskBackend *backend, BriskItem *item)
-{
-        g_assert(backend != NULL);
-        BriskBackendClass *klazz = BRISK_BACKEND_GET_CLASS(backend);
-
-        g_return_val_if_fail(klazz->unpin_item != NULL, FALSE);
-        return klazz->unpin_item(backend, item);
+        if (klazz->get_item_actions == NULL) {
```
Comment (ikeydoherty) [inline src/backend/backend.c]: For consistency could really be `g_return_val_if_fail(klazz->get_item_actions != NULL, NULL)`

File: src/lib/key-binder.c (hunk 7 above)
Comment (cybre) [inline src/lib/key-binder.c]: Yep

**Comment by ikeydoherty** [comment]
OK so I'm going to merge this now, on the premise that I'll optimise it after I've merged it. Letting @flexiondotorg know that @cybre has implemented the support for issue #8 and I'm happy for him to be awarded the bounty.

**Comment by ikeydoherty** [comment]
Outside of that, thank you @cybre for implementing this, it'll make a lot of folks happy. And it's about time we kicked up a gear on Brisk :)

**Pull request merged**